### PR TITLE
feat: clone and fetch work with empty remote repositories

### LINKED ISSUE #647: Throw dedicated error for empty remote repository
When the remote repository is empty, throw a dedicated error so an application can catch it. Currently, isomorphic-git throws an error for the first missing capability, which ends up being rather arbitrary and not very informative.

I've worked out that an empty repository will report no refs and no capabilities. We could check if either (or both) of those properties is empty to determine that the repository is empty.

### PR DESCRIPTION
**Pull request by mojavelinux**
resolves #647

### CODE REVIEW

**Comment by karma-pr-reporter** [comment]
## Test Results for 86a1c04c124acb06051f8539421dfc3db796903d:
| Browser | Passed | Skipped | Failed | Time | Disconnected | Error |
|---------|--------|---------|--------|------|--------------|-------|
| Firefox 63.0.0 (Ubuntu 0.0.0) | 77 | 1 | 0 | 43.945 secs | false | undefined |
| HeadlessChrome 0.0.0 (Linux 0.0.0) | 66 | 1 | 0 | 49.339 secs | false | undefined |
| Safari 12.0.0 (Mac OS X 10.13.6) | 17 | 1 | 0 | 1 min 18.005 secs | false | undefined |
| Chrome Mobile 67.0.3396 (Android 7.1.1) | 59 | 1 | 0 | 1 min 35.592 secs | false | undefined |
| Edge 17.17134.0 (Windows 10 0.0.0) | 115 | 1 | 0 | 2 mins 39.764 secs | false | undefined |
| Mobile Safari 11.0.0 (iOS 11.2.0) | 9 | 1 | 0 | 3 mins 52.612 secs | false | undefined |

**Comment by karma-pr-reporter** [comment]
## Test Results for 86a1c04c124acb06051f8539421dfc3db796903d:
| Browser | Passed | Skipped | Failed | Time | Disconnected | Error |
|---------|--------|---------|--------|------|--------------|-------|
| HeadlessChrome 0.0.0 (Linux 0.0.0) | 21 | 1 | 0 | 35.689 secs | false | undefined |
| Firefox 63.0.0 (Ubuntu 0.0.0) | 58 | 1 | 0 | 51.323 secs | false | undefined |
| Safari 12.0.0 (Mac OS X 10.13.6) | 104 | 1 | 0 | 1 min 51.36 secs | false | undefined |
| Edge 17.17134.0 (Windows 10 0.0.0) | 4 | 1 | 0 | 2 mins 8.785 secs | false | undefined |
| Chrome Mobile 67.0.3396 (Android 7.1.1) | 216 | 1 | 0 | 2 mins 54.287 secs | false | undefined |
| Mobile Safari 11.0.0 (iOS 11.2.0) | 60 | 1 | 0 | 3 mins 15.84 secs | false | undefined |

File: __tests__/test-clone.js [hunk 1]
```diff
@@ -112,4 +112,22 @@ describe('clone', () => {
     )
     expect(fs.existsSync(`${dir}/a.txt`)).toBe(true, `'a.txt' exists`)
   })
+
+  it.only('errors if remote repository is empty', async () => {
```
Comment (wmhilton) [inline __tests__/test-clone.js]: 😆 That's twice now. You just really want your tests to be the only ones, don't you!

File: src/commands/fetch.js [hunk 2]
```diff
@@ -175,6 +175,9 @@ async function fetchPackfile ({
     headers
   })
   auth = remoteHTTP.auth // hack to get new credentials from CredentialManager API
+  if (!remoteHTTP.capabilities.size) {
```
Comment (wmhilton) [inline src/commands/fetch.js]: It looks too much like `capabilities` is a regular Object and we're checking for a `size` capability on the server. Let's make this explicit:
```
// You cannot clone an empty repository.
if (remoteHTTP.capabilities.size === 0) {
```

hang on. Why are we looking at `remoteHTTP.capabilities`? Shouldn't we be looking at `remoteHTTP.refs`?

File: __tests__/test-clone.js (hunk 1 above)
Comment (mojavelinux) [reply __tests__/test-clone.js]: :man_facepalming:

File: src/commands/fetch.js (hunk 2 above)
Comment (mojavelinux) [reply src/commands/fetch.js]: That was exactly my question...and you've answered it. Both are empty in this condition, so I wasn't sure which one to rely on. But I agree refs would be more clear.

**Comment by karma-pr-reporter** [comment]
## Test Results for 98482bcda56cc86ad15405d53b47360f4564f8b9:
| Browser | Passed | Skipped | Failed | Time | Disconnected | Error |
|---------|--------|---------|--------|------|--------------|-------|
| HeadlessChrome 0.0.0 (Linux 0.0.0) | 23 | 1 | 0 | 37.137 secs | false | undefined |
| Firefox 63.0.0 (Ubuntu 0.0.0) | 67 | 1 | 0 | 1 min 7.53 secs | false | undefined |
| Safari 12.0.0 (Mac OS X 10.13.6) | 44 | 1 | 0 | 1 min 46.394 secs | false | undefined |
| Edge 17.17134.0 (Windows 10 0.0.0) | 91 | 1 | 0 | 2 mins 6.755 secs | false | undefined |
| Chrome Mobile 67.0.3396 (Android 7.1.1) | 88 | 1 | 0 | 3 mins 25.909 secs | false | undefined |
| Mobile Safari 11.0.0 (iOS 11.2.0) | 136 | 1 | 0 | 4 mins 25.772 secs | false | undefined |

**Comment by karma-pr-reporter** [comment]
## Test Results for 98482bcda56cc86ad15405d53b47360f4564f8b9:
| Browser | Passed | Skipped | Failed | Time | Disconnected | Error |
|---------|--------|---------|--------|------|--------------|-------|
| HeadlessChrome 0.0.0 (Linux 0.0.0) | 39 | 1 | 0 | 42.304 secs | false | undefined |
| Firefox 63.0.0 (Ubuntu 0.0.0) | 24 | 1 | 0 | 42.374 secs | false | undefined |
| Safari 12.0.0 (Mac OS X 10.13.6) | 147 | 1 | 0 | 2 mins 16.098 secs | false | undefined |
| Chrome Mobile 67.0.3396 (Android 7.1.1) | 18 | 1 | 0 | 3 mins 12.658 secs | false | undefined |
| Mobile Safari 11.0.0 (iOS 11.2.0) | 168 | 1 | 1 | 4 mins 29.2 secs | false | undefined |
| Edge 17.17134.0 (Windows 10 0.0.0) | 211 | 1 | 0 | 5 mins 15.781 secs | false | undefined |

**Comment by karma-pr-reporter** [comment]
## Test Results for 98482bcda56cc86ad15405d53b47360f4564f8b9:
| Browser | Passed | Skipped | Failed | Time | Disconnected | Error |
|---------|--------|---------|--------|------|--------------|-------|
| HeadlessChrome 0.0.0 (Linux 0.0.0) | 39 | 1 | 0 | 35.924 secs | false | undefined |
| Firefox 63.0.0 (Ubuntu 0.0.0) | 196 | 1 | 0 | 1 min 5.68 secs | false | undefined |
| Safari 12.0.0 (Mac OS X 10.13.6) | 127 | 1 | 0 | 1 min 52.209 secs | false | undefined |
| Chrome Mobile 67.0.3396 (Android 7.1.1) | 140 | 1 | 0 | 2 mins 27.734 secs | false | undefined |
| Edge 17.17134.0 (Windows 10 0.0.0) | 155 | 1 | 0 | 2 mins 58.648 secs | false | undefined |
| Mobile Safari 11.0.0 (iOS 11.2.0) | 52 | 1 | 0 | 3 mins 52.112 secs | false | undefined |

**Comment by wmhilton** [comment]
Getting there! I can clean it up from here. (We can't use sync `fs` operations in the tests because the BrowserFS HTTPRequest backend only supports async `fs` operations.)

**Comment by karma-pr-reporter** [comment]
## Test Results for 182eb99826c6d60439f9f95301fe2d9046a1790e:
| Browser | Passed | Skipped | Failed | Time | Disconnected | Error |
|---------|--------|---------|--------|------|--------------|-------|
| HeadlessChrome 0.0.0 (Linux 0.0.0) | 243 | 1 | 0 | 1 min 9.321 secs | false | undefined |
| Firefox 63.0.0 (Ubuntu 0.0.0) | 243 | 1 | 0 | 2 mins 8.103 secs | false | undefined |
| Safari 12.0.0 (Mac OS X 10.13.6) | 243 | 1 | 0 | 3 mins 21.248 secs | false | undefined |
| Mobile Safari 11.0.0 (iOS 11.2.0) | 243 | 1 | 0 | 4 mins 49.869 secs | false | undefined |
| Edge 17.17134.0 (Windows 10 0.0.0) | 243 | 1 | 0 | 5 mins 10.255 secs | false | undefined |
| Chrome Mobile 67.0.3396 (Android 7.1.1) | 243 | 1 | 0 | 7 mins 29.843 secs | false | undefined |

**Comment by mojavelinux** [comment]
:+1:

**Comment by wmhilton** [comment]
cc @pyramation

**Pull request merged**